src/profiling/profiler.hpp

```diff
@@ -1,255 +1,65 @@
 
-class profiler {
-public:
-    profiler(std::string name) :
-        root_region_(std::move(name))
-    {}
-
-    // the copy constructor doesn't do a "deep copy"
-    // it simply creates a new profiler with the same name
-    // This is needed for tbb to create a list of thread local profilers
-    profiler(const profiler& other) :
-        profiler(other.root_region_.name())
-    {}
-
-    /// step down into level with name
-    void enter(const char* name);
-
-    /// step up one level
-    void leave();
-
-    /// step up multiple n levels in one call
-    void leave(int n);
-
-    /// return a reference to the root region
-    region_type& regions() { return root_region_; }
-
-    /// return a pointer to the current region
-    region_type* current_region() { return current_region_; }
-
-    /// return if in the root region (i.e. the highest level)
-    bool is_in_root() const { return &root_region_ == current_region_; }
-
-    /// return if the profiler has been activated
-    bool is_activated() const { return activated_; }
-
-    /// start (activate) the profiler
-    void start();
-
-    /// stop (deactivate) the profiler
-    void stop();
-
-    /// restart the profiler
-    /// remove all trace information and restart timer for the root region
-    void restart();
-
-    /// the time stamp at which the profiler was started (avtivated)
-    timer_type::time_point start_time() const { return start_time_; }
-
-    /// the time stamp at which the profiler was stopped (deavtivated)
-    timer_type::time_point stop_time()  const { return stop_time_; }
-
-    /// the time in seconds between activation and deactivation of the profiler
-    double wall_time() const {
-        return timer_type::difference(start_time_, stop_time_);
-    }
-
-    /// stop the profiler then generate the performance tree ready for output
-    profiler_node performance_tree();
+void profiler_clear();
+void profiler_enter(std::size_t region_id);
+void profiler_leave();
 
-private:
-    void activate()   { activated_ = true;  }
-    void deactivate() { activated_ = false; }
+profile profiler_summary();
+std::size_t profiler_region_id(const char* name);
 
-    timer_type::time_point start_time_;
-    timer_type::time_point stop_time_;
-    bool activated_ = false;
-    region_type root_region_;
-    region_type* current_region_ = &root_region_;
-};
+std::ostream& operator<<(std::ostream&, const profile&);
 
 #ifdef ARB_HAVE_PROFILING
-namespace data {
-    using profiler_wrapper = arb::threading::enumerable_thread_specific<profiler>;
-    extern profiler_wrapper profilers_;
-}
-#endif
 
-/// get a reference to the thread private profiler
-/// will lazily create and start the profiler it it has not already been done so
-profiler& get_profiler();
+    // enter a profiling region
+    #define REGION_TAG_NAME(x) x##_profile_region_tag__
+    #define PE(name) \
+        { \
+            static std::size_t REGION_TAG_NAME(name) = arb::util::profiler_region_id(#name); \
```

Comment:
With the new block-based wrapper (which is very cool IMO!) there is no longer a need to use `REGION_TAG_NAME` to derive a unique identifier. Allows also change to e.g. '/' as default separator.

Reply:
Done.